model/interactive/src/main/proto/beam_interactive_api.proto

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Protocol Buffers describing a service that can be used in conjunction with
+ * the TestStream class in order to control a pipeline remotely.
+ */
+
+syntax = "proto3";
+
+package org.apache.beam.model.interactive.v1;
+
+option go_package = "interactive_v1";
+option java_package = "org.apache.beam.model.interactive.v1";
+option java_outer_classname = "BeamInteractiveApi";
+
+import "beam_runner_api.proto";
+import "google/protobuf/timestamp.proto";
+
+
+service InteractiveService {
+  // A TestStream will request for events using this RPC.
+  rpc Events(EventsRequest) returns (stream EventsResponse) {}
+}
+
+message EventsRequest { }
+message EventsResponse {
+  // The TestStreamPayloads that will be sent to the TestStream.
+  repeated org.apache.beam.model.pipeline.v1.TestStreamPayload.Event events = 1;
+
+  // Is true when there are no more events to read.
+  bool end_of_stream = 2;
+}
+
+// The first record to be read in an interactive stream. This contains metadata
+// about the stream and how to properly process it.
+message InteractiveStreamHeader {
+  // The PCollection tag this stream is associated with.
+  string tag = 1;
+}
+
+// A record is a recorded element that sound source produced. Its function is
+// to give enough information to the InteractiveService to create a faithful
+// recreation of the original source of data.
+message InteractiveStreamRecord {
```

Comment:
I don't fully understand what you are saying, but I'm guessing you want a single proto with a repeated field for events and watermark/clock advances? In that case, it wouldn't work. This is going to be the file format for storing the events of an unbounded source. The expectation is that the file may not fit into memory, so it will be streamed in one-by-one. This necessitates having individual records instead of a single proto for the file.

Reply:
It's not really isomorphic. The InteractiveStreamRecord has the added processing_time field. This is the absolute wall time seen at when that element/watermark event was recorded. The TestStreamPaylaod.AdvanceProcessingTime event only records deltas between two timestamps. It's also good practice to separate the proto formats for storage and in-memory use.

Reply:
Given a start time, they are isomorphic. I'm not sold on the value having a redundant format for use on disk vs. in memory, but if we go this route the file-based ones are really test specific even if the service isn't.